flycheck.el

```diff
@@ -2725,54 +2770,95 @@ current syntax check."
+(defun flycheck-may-check-automatically (&rest conditions)
+  "Determine whether the buffer may be checked under one of CONDITIONS.
 
 Read-only buffers may never be checked automatically.
 
-If CONDITION is non-nil, determine whether syntax may checked
-automatically according to
+If CONDITIONS are given, determine whether syntax may be checked
+under at least one of them, according to
 `flycheck-check-syntax-automatically'."
   (and (not (or buffer-read-only (flycheck-ephemeral-buffer-p)))
        (file-exists-p default-directory)
-       (or (not condition)
-           (memq condition flycheck-check-syntax-automatically))))
+       (or (not conditions)
+           (seq-some
+            (lambda (condition)
+              (memq condition flycheck-check-syntax-automatically))
+            conditions))))
+
+(defvar-local flycheck--idle-trigger-timer nil
+  "Timer used to trigger a syntax check after an idle delay.")
+
+(defvar-local flycheck--idle-trigger-conditions nil
+  "List of conditions under which an idle syntax check will be triggered.
+This will be some subset of the allowable values for
+`flycheck-check-syntax-automatically'.
+
+For example, if the user switches to a buffer and then makes an
+edit, this list will have the values `idle-change' and
+`idle-buffer-switch' in it, at least until the idle timer
+expires.")
 
 (defun flycheck-buffer-automatically (&optional condition force-deferred)
   "Automatically check syntax at CONDITION.
 
 Syntax is not checked if `flycheck-may-check-automatically'
-returns nil for CONDITION.
+returns nil for CONDITION.  (CONDITION may be a single condition
+or a list of them.)
 
 The syntax check is deferred if FORCE-DEFERRED is non-nil, or if
 `flycheck-must-defer-check' returns t."
-  (when (and flycheck-mode (flycheck-may-check-automatically condition))
+  (when (and flycheck-mode (if (listp condition)
+                               (apply #'flycheck-may-check-automatically
+                                      condition)
+                             (flycheck-may-check-automatically condition)))
+    (flycheck--clear-idle-trigger-timer)
+    (setq flycheck--idle-trigger-conditions nil)
     (if (or force-deferred (flycheck-must-defer-check))
         (flycheck-buffer-deferred)
       (with-demoted-errors "Error while checking syntax automatically: %S"
         (flycheck-buffer)))))
 
-(defvar-local flycheck-idle-change-timer nil
-  "Timer to mark the idle time since the last change.")
+(defun flycheck--clear-idle-trigger-timer ()
+  "Clear the idle trigger timer."
+  (when flycheck--idle-trigger-timer
+    (cancel-timer flycheck--idle-trigger-timer)
+    (setq flycheck--idle-trigger-timer nil)))
 
-(defun flycheck-clear-idle-change-timer ()
-  "Clear the idle change timer."
-  (when flycheck-idle-change-timer
-    (cancel-timer flycheck-idle-change-timer)
-    (setq flycheck-idle-change-timer nil)))
+(defun flycheck--handle-idle-trigger (buffer)
```

Comment:
I'm not sure whether we can do this: many people seem to overwrite this function in their config. https://github.com/search?q=flycheck-clear-idle-change-timer&type=Code

We should try to understand why they do that, and make sure we're not breaking their config.

Reply:
Ah, thanks, I remember.  There were two issues here, I think: @magnars' config, and the advice added by auto-complete.   Could we define an alias so that the advice still works?

Reply:
+1 for the `defalias`.  Seems a good way to maintain backward compatibility while keeping the internal code clean.

Reply:
Sure, I'll add an alias.